### PR TITLE
"Normalization"

### PR DESCRIPTION
**Pull request by bh1995**
Hey,
So I kind of don't think this is complete yet, but is this the right idea for the normalization? I also am not sure how to test this, do you have any ideas on how to test if the normalization is correct?
All the best, Björn.

### CODE REVIEW

File: gcn_prot/models/layers.py [hunk 1]
```diff
@@ -11,30 +11,41 @@ class GraphConvolution(nn.Module):
...
+        #     v = torch.cat([matrix for matrix in v])
+        # # apply each adj_i to each aminoacid_i
+        # support = torch.spmm(adj, v)
+        # # return to tensor
+        # support = support.reshape(v_shape)
+        # output = self.act(self.linear(support))
+        # output = F.dropout(output, training=self.training)
+        # return output, adj
+           c1 = self.weight1(v)
+           c2 = self.weight(v)
```
Comment (carrascomj) [inline gcn_prot/models/layers.py]: Typo here, I guess `self.weight2`

**Review by carrascomj — COMMENTED** [review]
You should not modify the GraphConvolution object, but create another new object (like `NormallizationLayer` or something like, in the same `layers.py` file ). It seems good to me, I can try to write a simple test for it and see if it works.

File: gcn_prot/models/layers.py [hunk 2]
```diff
@@ -41,3 +41,31 @@ def __repr__(self):
...
+        # Define trainable parameters
+        self.weight1 = nn.Linear(in_features, self.weight_feat, bias)
+        self.weight2 = nn.Linear(in_features, self.weight_feat, bias)
+        self.d = D
+
+    def forward(self, input):
+        """Pass forward features `v` and sparse adjacency matrix `adj`."""
+        v, adj = input
+        v_shape = v.shape
+        if len(v_shape) > 2:
```
Comment (carrascomj) [inline gcn_prot/models/layers.py:66]: This was done in the `GraphConvolution` to flatten the feature matrix `v` over the batch axis when there was effectively a batch dimension. I don't think that this will be needed here.

**Review by carrascomj — COMMENTED** [review]
I will add a test for it.

File: gcn_prot/models/layers.py (hunk 2 above)
Comment (carrascomj) [reply gcn_prot/models/layers.py:67]: It is now reflected in the tests that the online learning (no batch dimension) should also work.

File: gcn_prot/models/layers.py [hunk 3]
```diff
@@ -41,3 +41,32 @@ def __repr__(self):
...
+    def forward(self, input):
+        """Normalize sparse adjacency matrix `adj` in terms of `v`."""
+        v, adj = input
+        v_shape = v.shape
+        if len(v_shape) > 2:
+            c1 = self.weight1(v)
+            c2 = self.weight2(v)
+            c = (F.sigmoid(c1 + c2) * self.d) + 0.00001  # As to not divide by zero
+            norm_adj = torch.exp(-((adj * adj) / (2 * c * c)))
+            return norm_adj
```
Comment (carrascomj) [inline gcn_prot/models/layers.py:72]: The forward pass should retrieve both the unchanged `v` and `norm_adj`, because it need to pass both to the subsequent layers.

**Review by carrascomj — CHANGES_REQUESTED** [review]
The tests are not passing right now for some of the reasons that I have pointed out in the comments.

**Review by carrascomj — APPROVED** [review]
